File: core/src/test/java/com/pipai/adv/backend/battle/engine/KoTest.java
package com.pipai.adv.backend.battle.engine;

import com.pipai.adv.backend.battle.domain.*;
import com.pipai.adv.backend.battle.engine.commands.DevHpChangeCommand;
import com.pipai.adv.backend.battle.engine.log.BattleLogEvent;
import com.pipai.adv.backend.battle.engine.log.DamageEvent;
import com.pipai.adv.backend.battle.engine.log.NpcKoEvent;
import com.pipai.adv.backend.battle.engine.log.PlayerKoEvent;
import com.pipai.adv.npc.Npc;
import com.pipai.adv.npc.NpcList;
import com.pipai.adv.save.AdvSave;
import com.pipai.test.fixtures.TestFixturesKt;
import org.junit.Assert;
import org.junit.Test;

import java.util.List;

public class KoTest {

    @Test
    public void testPlayerKo() {
        AdvSave save = new AdvSave();
        NpcList npcList = new NpcList();
        BattleMap map = BattleMap.Factory.createBattleMap(4, 4);
        Npc player = TestFixturesKt.npcFromStats(new UnitStats(100, 1, 1, 1, 1, 1, 1, 1, 3),
                null);
        int playerId = npcList.addNpc(player);
        Npc enemy = TestFixturesKt.npcFromStats(new UnitStats(100, 1, 1, 1, 1, 1, 1, 1, 3),
                null);
        int enemyId = npcList.addNpc(enemy);
        map.getCell(0, 0).setFullEnvObject(new FullEnvObject.NpcEnvObject(playerId, Team.PLAYER, EnvObjTilesetMetadata.NONE));
        map.getCell(1, 1).setFullEnvObject(new FullEnvObject.NpcEnvObject(enemyId, Team.AI, EnvObjTilesetMetadata.NONE));

        BattleBackend backend = new BattleBackend(save, npcList, map);

        DevHpChangeCommand cmd = new DevHpChangeCommand(playerId, 0);

        Assert.assertTrue(backend.canBeExecuted(cmd).getExecutable());

        List<BattleLogEvent> events = backend.execute(cmd);
        Assert.assertTrue(events.stream().anyMatch(it -> it instanceof DamageEvent));
        DamageEvent damageEvent = (DamageEvent) events.stream().filter(it -> it instanceof DamageEvent).findFirst().get();
        Assert.assertEquals(playerId, damageEvent.getNpcId());
        Assert.assertEquals(0, player.getUnitInstance().getHp());

        Assert.assertTrue(events.stream().anyMatch(it -> it instanceof PlayerKoEvent));
        PlayerKoEvent koEvent = (PlayerKoEvent) events.stream().filter(it -> it instanceof PlayerKoEvent).findFirst().get();
        Assert.assertEquals(playerId, koEvent.getNpcId());

        FullEnvObject envObj = map.getCell(0, 0).getFullEnvObject();
        Assert.assertTrue(envObj instanceof FullEnvObject.NpcEnvObject);
        Assert.assertEquals(playerId, ((FullEnvObject.NpcEnvObject) envObj).getNpcId());
    }

    @Test
    public void testEnemyKo() {
        AdvSave save = new AdvSave();
        NpcList npcList = new NpcList();
        BattleMap map = BattleMap.Factory.createBattleMap(4, 4);
        Npc player = TestFixturesKt.npcFromStats(new UnitStats(100, 1, 1, 1, 1, 1, 1, 1, 3),
                null);
        int playerId = npcList.addNpc(player);
        Npc enemy = TestFixturesKt.npcFromStats(new UnitStats(100, 1, 1, 1, 1, 1, 1, 1, 3),
                null);
        int enemyId = npcList.addNpc(enemy);
        map.getCell(0, 0).setFullEnvObject(new FullEnvObject.NpcEnvObject(playerId, Team.PLAYER, EnvObjTilesetMetadata.NONE));
        map.getCell(1, 1).setFullEnvObject(new FullEnvObject.NpcEnvObject(enemyId, Team.AI, EnvObjTilesetMetadata.NONE));

        BattleBackend backend = new BattleBackend(save, npcList, map);

        DevHpChangeCommand cmd = new DevHpChangeCommand(enemyId, 0);

        Assert.assertTrue(backend.canBeExecuted(cmd).getExecutable());

        List<BattleLogEvent> events = backend.execute(cmd);
        Assert.assertTrue(events.stream().anyMatch(it -> it instanceof DamageEvent));
        DamageEvent damageEvent = (DamageEvent) events.stream().filter(it -> it instanceof DamageEvent).findFirst().get();
        Assert.assertEquals(enemyId, damageEvent.getNpcId());
        Assert.assertEquals(0, enemy.getUnitInstance().getHp());

        Assert.assertTrue(events.stream().anyMatch(it -> it instanceof NpcKoEvent));
        NpcKoEvent koEvent = (NpcKoEvent) events.stream().filter(it -> it instanceof NpcKoEvent).findFirst().get();
        Assert.assertEquals(enemyId, koEvent.getNpcId());
        Assert.assertEquals(null, map.getCell(1, 1).getFullEnvObject());
    }
}

File: core/src/test/java/com/pipai/adv/backend/battle/engine/KoTest.kt
package com.pipai.adv.backend.battle.engine

import com.pipai.adv.backend.battle.domain.*
import com.pipai.adv.backend.battle.engine.commands.DevHpChangeCommand
import com.pipai.adv.backend.battle.engine.log.DamageEvent
import com.pipai.adv.backend.battle.engine.log.NpcKoEvent
import com.pipai.adv.backend.battle.engine.log.PlayerKoEvent
import com.pipai.adv.domain.Npc
import com.pipai.adv.utils.AutoIncrementIdMap
import com.pipai.adv.utils.fetch
import com.pipai.test.fixtures.npcFromStats
import com.pipai.test.libgdx.GdxMockedTest
import com.pipai.test.libgdx.generateBackend
import org.junit.Assert
import org.junit.Test

class KoTest : GdxMockedTest() {

    @Test
    fun testPlayerKo() {
        val npcList = AutoIncrementIdMap<Npc>()
        val envObjList = AutoIncrementIdMap<EnvObject>()

        val map = BattleMap.createBattleMap(4, 4)
        val player = npcFromStats(UnitStats(100, 1, 1, 1, 1, 1, 1, 1, 3), null)
        val playerId = npcList.add(player)
        val enemy = npcFromStats(UnitStats(100, 1, 1, 1, 1, 1, 1, 1, 3), null)
        val enemyId = npcList.add(enemy)
        map.getCell(0, 0).fullEnvObjId = envObjList.add(NpcEnvObject(playerId, Team.PLAYER, EnvObjTilesetMetadata.NONE))
        map.getCell(1, 1).fullEnvObjId = envObjList.add(NpcEnvObject(enemyId, Team.AI, EnvObjTilesetMetadata.NONE))

        val backend = generateBackend(npcList, envObjList, map)

        val cmd = DevHpChangeCommand(playerId, 0)

        Assert.assertTrue(backend.canBeExecuted(cmd).executable)

        val events = backend.execute(cmd)
        Assert.assertTrue(events.stream().anyMatch { it -> it is DamageEvent })
        val (npcId) = events.stream().filter { it -> it is DamageEvent }.findFirst().get() as DamageEvent
        Assert.assertEquals(playerId.toLong(), npcId.toLong())
        Assert.assertEquals(0, player.unitInstance.hp.toLong())

        Assert.assertTrue(events.stream().anyMatch { it -> it is PlayerKoEvent })
        val (npcId1) = events.stream().filter { it -> it is PlayerKoEvent }.findFirst().get() as PlayerKoEvent
        Assert.assertEquals(playerId.toLong(), npcId1.toLong())

        val envObj = map.getCell(0, 0).fullEnvObjId
        Assert.assertTrue(envObj.fetch(envObjList) is NpcEnvObject)
        Assert.assertEquals(playerId.toLong(), (envObj.fetch(envObjList) as NpcEnvObject).npcId.toLong())
    }

    @Test
    fun testEnemyKo() {
        val npcList = AutoIncrementIdMap<Npc>()
        val envObjList = AutoIncrementIdMap<EnvObject>()

        val map = BattleMap.createBattleMap(4, 4)
        val player = npcFromStats(UnitStats(100, 1, 1, 1, 1, 1, 1, 1, 3), null)
        val playerId = npcList.add(player)
        val enemy = npcFromStats(UnitStats(100, 1, 1, 1, 1, 1, 1, 1, 3), null)
        val enemyId = npcList.add(enemy)
        map.getCell(0, 0).fullEnvObjId = envObjList.add(NpcEnvObject(playerId, Team.PLAYER, EnvObjTilesetMetadata.NONE))
        map.getCell(1, 1).fullEnvObjId = envObjList.add(NpcEnvObject(enemyId, Team.AI, EnvObjTilesetMetadata.NONE))

        val backend = generateBackend(npcList, envObjList, map)

        val cmd = DevHpChangeCommand(enemyId, 0)

        Assert.assertTrue(backend.canBeExecuted(cmd).executable)

        val events = backend.execute(cmd)
        Assert.assertTrue(events.stream().anyMatch { it -> it is DamageEvent })
        val (npcId) = events.stream().filter { it -> it is DamageEvent }.findFirst().get() as DamageEvent
        Assert.assertEquals(enemyId.toLong(), npcId.toLong())
        Assert.assertEquals(0, enemy.unitInstance.hp.toLong())

        Assert.assertTrue(events.stream().anyMatch { it -> it is NpcKoEvent })
        val (npcId1) = events.stream().filter { it -> it is NpcKoEvent }.findFirst().get() as NpcKoEvent
        Assert.assertEquals(enemyId.toLong(), npcId1.toLong())
        Assert.assertEquals(null, map.getCell(1, 1).fullEnvObjId)
    }
}
